src/charm.py

```diff
@@ -0,0 +1,323 @@
+    def _on_master_changed(self, event: pgsql.MasterChangedEvent) -> None:
+        """Handle changes in the primary database unit.
+
+        Args:
+            event: Event triggering the database master changed handler.
+        """
+        if event.master is None:
+            self._stored.db_name = None
+            self._stored.db_user = None
+            self._stored.db_password = None
+            self._stored.db_host = None
+            self._stored.db_port = None
+        else:
+            self._stored.db_name = event.master.dbname
+            self._stored.db_user = event.master.user
+            self._stored.db_password = event.master.password
+            self._stored.db_host = event.master.host
+            self._stored.db_port = event.master.port
+
+        self._on_config_changed(event)
+
+    def _on_install(self, _):
+        """Generate a config template to be rendered later."""
+        self._run_generate_synapse()
+
+    @property
+    def _external_hostname(self):
+        """Return the external hostname passed to ingress via the relation."""
+        # It is recommended to default to `self.app.name` so that the external
+        # hostname will correspond to the deployed application name in the
+        # model, but allow it to be set to something specific via config.
+        return self.config["server_name"] or self.app.name
+
+    def _container(self):
+        """Get the Synapse workload container.
+
+        Returns:
+            The pebble instance of the Synapse container.
+        """
+        return self.unit.get_container(self._CONTAINER_NAME)
+
+    def _populate_synapse_env_settings(self):
+        """Populate env settings to pass around."""
+        if os.getenv("SYNAPSE_SERVER_NAME") is None:
+            os.environ["SYNAPSE_SERVER_NAME"] = self.config["server_name"]
+        if os.getenv("SYNAPSE_REPORT_STATS") is None:
+            os.environ["SYNAPSE_REPORT_STATS"] = self.config["report_stats"]
+        pod_config = {
+                "SYNAPSE_SERVER_NAME": os.getenv("SYNAPSE_SERVER_NAME"),
+                "SYNAPSE_REPORT_STATS": os.getenv("SYNAPSE_REPORT_STATS"),
+                "SYNAPSE_NO_TLS": "True",
+                "POSTGRES_DB": self._stored.db_name,
+                "POSTGRES_HOST": self._stored.db_host,
+                "POSTGRES_PORT": self._stored.db_port,
+                "POSTGRES_USER": self._stored.db_user,
+                "POSTGRES_PASSWORD": self._stored.db_password,
+                }
+        return pod_config
+
+    def _register_user(self, event):
+        """Registers a user for usage with Synapse."""
+        Result = collections.namedtuple(
+                "CommandExecResult", "return_code stdout stderr")
+
+        user = event.params["username"]
+        password = event.params["password"]
+        admin = event.params["admin"]
+        admin_switch = "--admin" if admin == "yes" else "--no-admin"
+
+        cmd = ["register_new_matrix_user",
```

Comment:
Would it be possible to use the SDK instead of commands? https://matrix.org/docs/projects/sdk/matrix-org-python-sdk

Reply:
I haven't looked into it, for the PoC I went with what is used and described in the synapse docs for the dockerized version. If there are more commands coming up, it makes sense to use an SDK. For now, I think it is just this one and it is tied to the PoC - I don't expect to create users via that action except maybe for admins?